src/huggingface_hub/snapshot_download.py

```diff
@@ -0,0 +1,62 @@
+import os
+from pathlib import Path
+from typing import Optional, Union
+
+from .constants import HUGGINGFACE_HUB_CACHE
+from .file_download import cached_download, hf_hub_url
+from .hf_api import HfApi
+
+
+REPO_ID_SEPARATOR = "__"
+# ^ make sure this substring is not allowed in repo_ids on hf.co
+
+
+def snapshot_download(
+    repo_id: str,
+    revision: Optional[str] = None,
+    cache_dir: Union[str, Path, None] = None,
+) -> str:
+    """
+    Downloads a whole snapshot of a repo's files at the specified revision.
+    This is useful when you want all files from a repo, because you don't know
+    which ones you will need a priori.
+    All files are nested inside a folder in order to keep their actual filename
+    relative to that folder.
+
+    An alternative would be to just clone a repo but this would require that
+    the user always has git and git-lfs installed, and properly configured.
+
+    Note: at some point maybe this format of storage should actually replace
+    the flat storage structure we've used so far (initially from allennlp
+    if I remember correctly).
+
+    Return:
+        Local folder path (string) of repo snapshot
+    """
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+    if isinstance(cache_dir, Path):
+        cache_dir = str(cache_dir)
+
+    _api = HfApi()
+    model_info = _api.model_info(repo_id=repo_id, revision=revision)
+
+    storage_folder = os.path.join(
+        cache_dir, repo_id.replace("/", REPO_ID_SEPARATOR) + "." + model_info.sha
+    )
+
```

Comment:
Maybe here it would be possible to directly return the `storage_folder` if it's already been downloaded ?

This way we can save all the HEAD calls to get the etags in `cached_download`.

Reply:
I'm not sure, because this increases the chance that a corrupted download is going to happen and things will fail silently.